add torch.float64 precision support to the transformer test suite in TP/SP

tianyu-l opened the pull request:
This PR (as a followup to #115530) resolves previous issues of not passing `assertEqual()` tests (with small error) when comparing outputs from the single-gpu model and the distributed model, under certain input/model sizes or when certain operations (e.g. weight-tying) are enabled. This is done by simply enabling higher precision computation using `dtype=torch.float64`.

What is not tested: whether or not distributed model training convergence rate is affected using just `torch.float32` precision.

Test plan:
TP: `python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_False`
TP+SP: `python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_True`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116436



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @wconstab @yf225

tianyu-l committed:
add torch.float64 precision support to the transformer test suite in TP/SP

[ghstack-poisoned]

wanchaol commented on torch/testing/_internal/distributed/_tensor/common_dtensor.py:
```diff
@@ -69,13 +69,14 @@ def reset_parameters(self):
-    max_seq_len: int = 16
-    dim: int = 8
+    vocab_size: int = 28
+    max_seq_len: int = 32
```
curious why we need to enlarge the sizes here?

tianyu-l replied:
These new parameters didn't work before we use double precision. Since we know they will work, I just changed back to old parameters to minimize change.

wanchaol commented on torch/testing/_internal/distributed/_tensor/common_dtensor.py:
```diff
@@ -86,11 +87,12 @@ def __init__(self, args: ModelArgs):
-        self.wk = nn.Linear(args.dim, args.dim, bias=False)
-        self.wv = nn.Linear(args.dim, args.dim, bias=False)
-        self.wo = nn.Linear(args.dim, args.dim, bias=False)
+        self.wq = nn.Linear(args.dim, args.dim, bias=False, dtype=args.dtype)
```
instead of passing dtype to every layer initialization, can we simply do:

`Transformer.to(dtype=args.dtype)`

instead?

tianyu-l replied:
Yeah good point, sorry I didn't know this API. In this case, I think it is unnecessary to pass `dtype` into `ModelArgs`.

tianyu-l committed:
Update on "add torch.float64 precision support to the transformer test suite in TP/SP"

This PR (as a followup to #115530) resolves previous issues of not passing `assertEqual()` tests (with small error) when comparing outputs from the single-gpu model and the distributed model, under certain input/model sizes or when certain operations (e.g. weight-tying) are enabled. This is done by simply enabling higher precision computation using `dtype=torch.float64`.

What is not tested: whether or not distributed model training convergence rate is affected using just `torch.float32` precision.

Test plan:
TP: `python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_False`
TP+SP: `python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_True`




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm, sorry for the late stamp!

awgu commented:
Using fp64 to check functional correctness sounds good, but I wanted to be a bit careful about the numeric issue and understand it better. It may not be a good sign for us if a small fp32 model/input can already drift compared to single-GPU (since people will be running fp32 or smaller bit-width in practice).

Interestingly, I cannot repro the `assertEqual` failures when using `weight_tying=True` and `inp_size = [8, 12]`. I will share one source of numeric issue I have seen in the past: single vs. multi-tensor Adam. Since `DTensor` does not currently default to multi-tensor Adam (requiring passing `foreach=True` explicitly), we can see numeric drift comparing the single-GPU non-`DTensor` and multi-GPU `DTensor` cases. However, this drift can be fixed by passing `foreach=True` for the `DTensor` model's optimizer. It might be worthwhile to check if this solves the issue without fp64.

tianyu-l commented:
@pytorchbot rebase

tianyu-l committed:
Update on "add torch.float64 precision support to the transformer test suite in TP/SP"

This PR (as a followup to #115530) resolves previous issues of not passing `assertEqual()` tests (with small error) when comparing outputs from the single-gpu model and the distributed model, under certain input/model sizes or when certain operations (e.g. weight-tying) are enabled. This is done by simply enabling higher precision computation using `dtype=torch.float64`.

What is not tested: whether or not distributed model training convergence rate is affected using just `torch.float32` precision.

Test plan:
TP: `python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_False`
TP+SP: `python test/distributed/tensor/parallel/test_tp_examples.py -k test_transformer_training_is_seq_parallel_True`




cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 wconstab yf225

[ghstack-poisoned]

tianyu-l commented:
@pytorchbot merge
